jupyterhub_traefik_proxy/toml.py

```diff
@@ -0,0 +1,275 @@
+- Route paths should be normalized to always start and end with '/'
+"""
+
+# Copyright (c) Jupyter Development Team.
+# Distributed under the terms of the Modified BSD License.
+
+import json
+import os
+import asyncio
+import toml
+
+from traitlets import Any, default, Unicode
+
+from . import traefik_utils
+from jupyterhub.proxy import Proxy
+from jupyterhub_traefik_proxy import TraefikProxy
+
+
+class TraefikTomlProxy(TraefikProxy):
+    """JupyterHub Proxy implementation using traefik and toml config file"""
+
+    mutex = Any()
+
+    @default("mutex")
+    def _default_mutex(self):
+        return asyncio.Lock()
+
+    toml_static_config_file = Unicode(
+        "traefik.toml", config=True, help="""traefik's static configuration file"""
+    )
+
+    toml_dynamic_config_file = Unicode(
+        "rules.toml", config=True, help="""traefik's dynamic configuration file"""
+    )
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.routes_cache = {}
+
+    async def _setup_traefik_static_config(self):
+        await super()._setup_traefik_static_config()
+
+        self.static_config["file"] = {"filename": "rules.toml", "watch": True}
+        toml_string = toml.dumps(self.static_config)
+
+        try:
+            with open(self.toml_static_config_file, "w") as f:
+                f.write(toml_string)
+            open(self.toml_dynamic_config_file, "a").close()
+        except IOError:
+            self.log.exception("Couldn't set up traefik's static config.")
+        except:
+            self.log.error(
+                "Couldn't set up traefik's static config. Unexpected error:",
+                sys.exc_info()[0],
+            )
+
+    def _start_traefik(self):
+        self.log.info("Starting traefik...")
+        try:
+            self._launch_traefik(config_type="toml")
+        except FileNotFoundError as e:
+            self.log.error(
+                "Failed to find traefik \n"
+                "The proxy can be downloaded from https://github.com/containous/traefik/releases/download."
+            )
+            raise
+
+    def _clean_resources(self):
+        try:
+            os.remove(self.toml_static_config_file)
+            os.remove(self.toml_dynamic_config_file)
+        except:
+            self.log.error("Failed to remove traefik's configuration files \n")
+            raise
+
+    def _get_route_unsafe(self, routespec):
+        try:
+            result = {
+                "routespec": routespec,
+                "target": self.routes_cache[routespec]["target"],
+                "data": self.routes_cache[routespec]["data"],
+            }
+        except KeyError:
+            self.log.info("No route for {} doesn't exist!".format(routespec))
+            result = None
+        finally:
+            return result
+
+    def _persist_routes(self):
+        with traefik_utils.atomic_writing(self.toml_dynamic_config_file) as config_fd:
+            config_fd.write("[frontends]\n")
```

Comment:
use `toml.dump` here?

Reply:
I agree that it doesn't look pretty as it is, but the `routes_cache` dict is structured so that we can easily delete and get routes from it. `toml.dump` won't work on the dict as it is. We could modify it to make sense for traefik and `toml.dump`, but the deletion and retrieving of routes would complicate a lot in my opinion. 
I could get rid of the tabs in `add_route` since they don't matter (I've noticed toml.dump doesn't print them) and make the `routes_cache` prettier.

Reply:
We need to be able to load state from this file, e.g. when jupyterhub restarts while the proxy is still running. This means that all data (target, routespec, data) must be retrievable from the toml. So in general, we need to have the ability to transform from this toml file to internal routes and back. For that reason, I don't think storing toml strings is the right thing to do. Plus, it is vulenerable to toml syntax errors, which writing toml data directly isn't.

Reply:
You're right. I didn't think of this case. Do you agree to add the routespec to the frontend/backend aliases in order to be able to find them easier? Will those be unique? Right now the alias is build based on the target.

Reply:
In fact, it's probably a good idea to isolate and test the roundtrip from the `get_routes` structure to/from toml to make sure it's stable.